### PR TITLE
Parameterize offline-storage deletes and harden kill-switch response handling

### PR DESCRIPTION
**Pull request by bmehta001**
Builds `OfflineStorage_SQLite::DeleteRecords` as a parameterized prepared statement and tightens kill-switch response handling.

## Changes
- `OfflineStorage_SQLite::DeleteRecords` now builds a **parameterized** statement: column names are taken from a fixed whitelist and values are bound via `sqlite3_bind_*` on a single prepared statement instead of being concatenated into the SQL text. This is more robust and avoids quoting/escaping edge cases for values such as tenant tokens. Prepare/bind/execute results are checked and logged (with `sqlite3_errmsg`) on failure, and a non-numeric value for an integer filter column aborts the delete instead of coercing to `0`.
- `DeleteRecords` no longer issues a `DELETE` with an empty predicate, and fails closed (deletes nothing) on an unrecognized filter column.
- `KillSwitchManager::handleResponse` rejects empty kill-token values, values containing control characters, and values longer than 256 bytes (`kMaxTenantTokenBytes`, a generous ceiling well above the ~74-char real token size) before storing them, while keeping opaque tokens — which may legitimately contain spaces/quotes — killable.
- `KillSwitchManager::handleResponse` parses the `Retry-After` and `kill-duration` response headers through a non-throwing helper that accepts only an RFC 7231 `1*DIGIT` value surrounded by HTTP OWS (SP/HTAB) and clamps the result to a safe maximum, so a malformed value (the HTTP-date form of `Retry-After`, a leading sign, stray CR/LF, or an overflow-inducing magnitude) is ignored rather than propagating out of the SDK worker thread or wrapping the computed expiry time.
- Adds `OfflineStorageTests_SQLite` regression tests for the parameterized `DeleteRecords`: deletion by `tenant_token` (including unusual characters and a SQL-injection attempt), deletion by a numeric column, multi-column `AND` filters, and fail-closed behavior on an empty filter, an invalid numeric value, and an unrecognized column.
- Adds `KillSwitchManagerTests` covering valid input plus malformed `Retry-After` / `kill-duration` values (non-numeric, HTTP-date, out-of-range, trailing/leading garbage, leading sign, huge/clamped), control-character, empty, and opaque kill-tokens, and `:all` suffix stripping. These tests now also build and run on the Android and iOS native test targets.

## Validation
- **`KillSwitchManagerTests` — 20 tests** (the full new suite): valid `Retry-After`/kill-token input; malformed `Retry-After`/`kill-duration` (non-numeric, HTTP-date, out-of-range, trailing garbage, trailing/leading OWS, trailing CR/LF, leading sign, huge-clamped); control-character, empty, and opaque kill-tokens; and `:all` suffix stripping (including a token that strips to empty); and the tenant-token length cap (a 256-byte token stays killable, a 257-byte token is ignored). Now built/run on the host unit tests, Android, and iOS.
- **`OfflineStorageTests_SQLite` — 32 tests total, of which 7 exercise the parameterized `DeleteRecords`**: SQL-injection attempt, delete-by-`tenant_token` (incl. unusual chars), delete-by-numeric-column, multi-column `AND`, empty filter, invalid numeric filter, and fail-closed on an unrecognized column.
- The core parameterized-delete subset was validated locally on Windows x64 Debug; the full suites (including the additions above) run in CI across the host unit-test platforms, with the kill-switch tests additionally covered on Android and iOS.

### CODE REVIEW

File: lib/offline/OfflineStorage_SQLite.cpp [hunk 1]
```diff
@@ -413,40 +412,70 @@ namespace MAT_NS_BEGIN {
...
-                        // string types
-                        quotes = true;
-                    } 
-                    else if (
-                        // integer types
-                        (kv.first == "latency") ||
-                        (kv.first == "persistence") ||
-                        (kv.first == "retry_count"))
+                    LOG_WARN("DeleteRecords: ignoring unrecognized filter column");
+                    continue;
```
[the remark this answers is not in the document]
Comment (bmehta001) [reply lib/offline/OfflineStorage_SQLite.cpp]: Done in 048709ce - the warning now includes the rejected column name: `ignoring unrecognized filter column '%s'`.

File: lib/offline/OfflineStorage_SQLite.cpp [hunk 2]
```diff
@@ -413,40 +412,70 @@ namespace MAT_NS_BEGIN {
...
+            SqliteStatement stmt(*m_db, sql.c_str());
+            int idx = 1;
+            for (const auto& it : boundColumns)
+            {
+                const std::string& value = it->second;
+                if (kAllowedColumns.at(it->first) == ColumnType::Text)
+                {
+                    g_sqlite3Proxy->sqlite3_bind_text(stmt.handle(), idx,
+                        value.data(), static_cast<int>(value.size()), SQLITE_STATIC);
+                }
```
[the remark this answers is not in the document]
Comment (bmehta001) [reply lib/offline/OfflineStorage_SQLite.cpp:474]: Done in 048709ce - `sqlite3_bind_*` return codes are now checked and the delete aborts (logging `sqlite3_errmsg`) on failure. A non-numeric value for an integer column is now treated as an invalid filter and aborts the delete (using `std::stoll` full-consumption check) instead of coercing to 0.

File: lib/offline/OfflineStorage_SQLite.cpp [hunk 3]
```diff
@@ -413,40 +412,70 @@ namespace MAT_NS_BEGIN {
...
-                        kv.second;
+                    g_sqlite3Proxy->sqlite3_bind_int64(stmt.handle(), idx, numeric);
                 }
-                return clause;
-            };
-            std::string sql = "DELETE FROM " TABLE_NAME_EVENTS " WHERE ";
-            Execute(sql + formatter(whereFilter));
+                ++idx;
+            }
+            stmt.execute();
```
[the remark this answers is not in the document]
Comment (bmehta001) [reply lib/offline/OfflineStorage_SQLite.cpp]: Done in 048709ce - the result of `execute()` is now checked and an error is logged when the DELETE fails.

File: lib/offline/KillSwitchManager.hpp [hunk 4]
```diff
@@ -156,6 +166,50 @@ namespace MAT_NS_BEGIN {
...
+        {
+            try
+            {
+                outSeconds = static_cast<int64_t>(std::stoll(value));
+                return true;
+            }
+            catch (const std::exception&)
+            {
+                return false;
+            }
```
[the remark this answers is not in the document]
Comment (bmehta001) [reply lib/offline/KillSwitchManager.hpp:239]: Done in 048709ce - `tryParseSeconds` now sets `outSeconds = 0` on parse failure.

File: lib/offline/KillSwitchManager.hpp [hunk 5]
```diff
@@ -156,6 +163,49 @@ namespace MAT_NS_BEGIN {
...
+            {
+                outSeconds = static_cast<int64_t>(std::stoll(value));
+                return true;
+            }
+            catch (const std::exception&)
+            {
+                outSeconds = 0;
+                return false;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (bmehta001) [reply lib/offline/KillSwitchManager.hpp:240]: Done in c7819031 - tryParseSeconds now checks full consumption (allowing only trailing whitespace) and returns false on trailing garbage like `120; foo`, so a malformed header is ignored rather than parsed from its numeric prefix. Added regression tests for the garbage and trailing-whitespace cases.

File: lib/offline/KillSwitchManager.hpp [hunk 6]
```diff
@@ -7,7 +7,9 @@
 
 #include "pal/PAL.hpp"
 
+#include <cctype>
 #include <map>
+#include <stdexcept>
 #include <string>
 #include <mutex>
```
[the remark this answers is not in the document]
Comment (bmehta001) [reply lib/offline/KillSwitchManager.hpp]: Done in 13aaeeb2 - added direct includes for `<vector>` and `<list>` (and tidied the include block) so the header is self-contained and not dependent on include order.

File: lib/offline/KillSwitchManager.hpp [hunk 7]
```diff
@@ -156,6 +166,61 @@ namespace MAT_NS_BEGIN {
...
+                return false;
+            }
+        }
+
+        // A tenant token is the API ingestion key tenant id: alphanumeric with
+        // '-', '_' or '.' separators. Reject anything outside that set.
+        static bool isValidTenantToken(const std::string& token)
+        {
+            if (token.empty() || token.size() > 256)
+            {
```
[the remark this answers is not in the document]
Comment (bmehta001) [reply lib/offline/KillSwitchManager.hpp]: Good catch - went with option (b) in 8e975353. Relaxed isValidTenantToken to reject only genuinely unsafe content (control characters incl. CR/LF for log-injection safety, plus the length cap) and accept any other byte. Tenant tokens are opaque (spaces/quotes are stored elsewhere) and the offline-storage DELETE is parameterized, so any value is safe to act on and stays killable. I did not add ingestion-time validation (option a) since the SDK intentionally treats tenantToken as opaque and that would risk rejecting tokens accepted today. Tests updated: a control-char token is rejected; an opaque quotes/spaces token remains blockable.

File: lib/offline/KillSwitchManager.hpp [hunk 8]
```diff
@@ -156,6 +167,62 @@ namespace MAT_NS_BEGIN {
...
+                // Reject trailing garbage (e.g. "120; foo"); only trailing
+                // whitespace is allowed, so a malformed header is ignored rather
+                // than silently parsed from its numeric prefix.
+                for (size_t i = consumed; i < value.size(); ++i)
+                {
+                    if (std::isspace(static_cast<unsigned char>(value[i])) == 0)
+                    {
+                        return false;
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (bmehta001) [reply lib/offline/KillSwitchManager.hpp]: Good catch - fixed in 986c9b44. `tryParseSeconds` now tolerates only HTTP OWS (SP / HTAB per RFC 7230) after the number instead of `std::isspace()`, so CR/LF (and form-feed/vertical-tab) are treated as malformed and the header is ignored. `"120\r\n"` is now rejected. Removed the now-unused `<cctype>` include and added a regression test `handleResponse_RetryAfterWithTrailingCRLF_IsIgnored`; the existing trailing-whitespace test already used spaces, so it still passes.

File: lib/offline/KillSwitchManager.hpp [hunk 9]
```diff
@@ -156,6 +166,65 @@ namespace MAT_NS_BEGIN {
...
+        // drives handleResponse has no exception guard, so a throw here would
+        // crash the process. A standards-compliant server may legitimately send
+        // Retry-After as an HTTP-date, which is non-numeric and must be ignored.
+        static bool tryParseSeconds(const std::string& value, int64_t& outSeconds)
+        {
+            outSeconds = 0;
+            try
+            {
+                size_t consumed = 0;
+                const long long parsed = std::stoll(value, &consumed);
```
[the remark this answers is not in the document]
Comment (bmehta001) [reply lib/offline/KillSwitchManager.hpp]: Agreed - fixed in b7abaeec. Rewrote `tryParseSeconds` to trim only HTTP OWS (SP / HTAB) from both ends and require the remainder to be `1*DIGIT` before calling `std::stoll`. This rejects the leading `'+'/'-'` sign and leading CR/LF that `std::stoll` would otherwise skip, making leading and trailing handling symmetric and conformant with RFC 7231 delay-seconds. Added tests for leading-OWS (accepted) and leading-sign (ignored); validated all edge cases (`"  120"`->120, `"+120"`->ignored, `"\r\n120"`->ignored, overflow->ignored).

File: lib/offline/OfflineStorage_SQLite.cpp [hunk 10]
```diff
@@ -413,40 +413,94 @@ namespace MAT_NS_BEGIN {
...
+            // Never run a DELETE with no predicate, which would erase the table.
+            if (clause.empty())
+            {
+                LOG_WARN("DeleteRecords: no recognized filter columns; nothing deleted");
+                return;
+            }
+
+            const std::string sql = "DELETE FROM " TABLE_NAME_EVENTS " WHERE " + clause;
+            SqliteStatement stmt(*m_db, sql.c_str());
+            int idx = 1;
```
[the remark this answers is not in the document]
Comment (bmehta001) [reply lib/offline/OfflineStorage_SQLite.cpp:465]: Fixed defensively in e2bf6229 - added an explicit `if (stmt.handle() == nullptr)` guard that logs `failed to prepare` and returns before binding. Note this was not actually a crash: `sqlite3_bind_*` on a null `sqlite3_stmt*` returns `SQLITE_MISUSE` (via SQLite's `vdbeSafetyNotNull`), which the existing per-bind `rc != SQLITE_OK` check already catches and aborts on. The explicit guard makes the intent obvious and gives a clearer error than a per-column MISUSE.

File: lib/offline/KillSwitchManager.hpp [hunk 11]
```diff
@@ -39,8 +42,8 @@ namespace MAT_NS_BEGIN {
             std::string timeStr = headers.get("Retry-After");
             if (!timeStr.empty())
             {
-                int64_t timeinSecs = std::stoi(timeStr);
-                if (timeinSecs > 0)
+                int64_t timeinSecs = 0;
+                if (tryParseSeconds(timeStr, timeinSecs) && timeinSecs > 0)
                 {
                     std::lock_guard<std::mutex> guard(m_lock);
                     m_retryAfterExpiryTime = PAL::getUtcSystemTime() + timeinSecs;
```
[the remark this answers is not in the document]
Comment (bmehta001) [reply lib/offline/KillSwitchManager.hpp:49]: Fixed in e2bf6229. Both `Retry-After` and `kill-duration` flow through `tryParseSeconds`, which now clamps its result to ~100 years (`100LL*365*24*60*60`, well below `INT64_MAX` and far above any legitimate value). `getUtcSystemTime() + timeinSecs` therefore can no longer overflow/wrap the expiry time. Added test `handleResponse_HugeRetryAfter_IsClampedNotWrapped` and validated that `now + clamped` stays well within range.

File: lib/offline/KillSwitchManager.hpp [hunk 12]
```diff
@@ -64,14 +67,21 @@ namespace MAT_NS_BEGIN {
...
 
                 int64_t timeinSecs = 0;
                 std::string timeString = headers.get("kill-duration");
                 if (!timeString.empty())
                 {
-                    timeinSecs = std::stoi(timeString);
+                    tryParseSeconds(timeString, timeinSecs);
                 }
 
                 if (killtokensVector.size() > 0 && timeinSecs > 0)
```
[the remark this answers is not in the document]
Comment (bmehta001) [reply lib/offline/KillSwitchManager.hpp:90]: Fixed in e2bf6229. Both `Retry-After` and `kill-duration` flow through `tryParseSeconds`, which now clamps its result to ~100 years (`100LL*365*24*60*60`, well below `INT64_MAX` and far above any legitimate value). `getUtcSystemTime() + timeinSecs` therefore can no longer overflow/wrap the expiry time. Added test `handleResponse_HugeRetryAfter_IsClampedNotWrapped` and validated that `now + clamped` stays well within range.

File: lib/offline/OfflineStorage_SQLite.cpp [hunk 13]
```diff
@@ -413,40 +413,99 @@ namespace MAT_NS_BEGIN {
...
+                LOG_WARN("DeleteRecords: no recognized filter columns; nothing deleted");
+                return;
+            }
+
+            const std::string sql = "DELETE FROM " TABLE_NAME_EVENTS " WHERE " + clause;
+            SqliteStatement stmt(*m_db, sql.c_str());
+            if (stmt.handle() == nullptr)
+            {
+                LOG_ERROR("DeleteRecords: failed to prepare delete statement for table " TABLE_NAME_EVENTS);
+                return;
```
[the remark this answers is not in the document]
Comment (bmehta001) [reply lib/offline/OfflineStorage_SQLite.cpp:463]: Fixed in d5c7706b - the prepare-failure path now logs `g_sqlite3Proxy->sqlite3_errmsg(*m_db)` like the bind/execute paths in the same function.

File: tests/unittests/KillSwitchManagerTests.cpp [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+using namespace Microsoft::Applications::Events;
+
+TEST(KillSwitchManagerTests, handleResponse_ValidRetryAfter_ActivatesRetryAfter)
+{
+    KillSwitchManager manager;
+    HttpHeaders headers;
+    headers.add("Retry-After", "120");
+    manager.handleResponse(headers);
+    ASSERT_TRUE(manager.isRetryAfterActive());
+}
```
[the remark this answers is not in the document]
Comment (bmehta001) [reply tests/unittests/KillSwitchManagerTests.cpp:17]: Updated the PR description to match the current tests: `OfflineStorageTests_SQLite` is 28 tests (3 new) and `KillSwitchManagerTests` is 14 tests, 42 total. (The file defines 14 `TEST` cases, not 15 - the extra edge-case tests were added while addressing earlier review rounds.)

File: lib/offline/KillSwitchManager.hpp [hunk 15]
```diff
@@ -156,6 +167,101 @@ namespace MAT_NS_BEGIN {
...
+        // offline-storage DELETE is parameterized (SQLite bind / Room DAO) and the
+        // in-memory match is a plain string compare -- so any printable value must
+        // remain killable; over-restricting would let that tenant escape the kill.
+        // We still reject control characters and over-long values as defensive
+        // hygiene: a legitimate token never contains them, an embedded NUL would be
+        // truncated by the Room JNI NewStringUTF(c_str()) call (acting on the wrong
+        // token), and it keeps the value safe if it ever reaches a log/display sink.
+        static bool isValidTenantToken(const std::string& token)
+        {
+            if (token.empty() || token.size() > 256)
```
Comment (lalitb) [inline lib/offline/KillSwitchManager.hpp]: Do we have an authoritative 256-byte max for tenant tokens? I don't see the same limit enforced when apps create loggers, so this could make a valid stored tenant impossible to kill if its token is longer than 256 bytes.

File: tests/unittests/CMakeLists.txt [hunk 16]
```diff
@@ -27,6 +27,7 @@ set(SRCS
   HttpResponseDecoderTests.cpp
   HttpServerTests.cpp
   InformationProviderImplTests.cpp
+  KillSwitchManagerTests.cpp
```
Comment (lalitb) [inline tests/unittests/CMakeLists.txt:30]: should we add this for Android and iOS native unit tests too ?

File: tests/unittests/KillSwitchManagerTests.cpp [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+    // RFC 7231 delay-seconds is 1*DIGIT; a leading '+'/'-' sign (which std::stoll
+    // would accept) is malformed and must be ignored.
+    KillSwitchManager manager;
+    HttpHeaders headers;
+    headers.add("Retry-After", "+120");
+    manager.handleResponse(headers);
+    ASSERT_FALSE(manager.isRetryAfterActive());
+}
+
+TEST(KillSwitchManagerTests, handleResponse_HugeRetryAfter_IsClampedNotWrapped)
```
Comment (lalitb) [inline tests/unittests/KillSwitchManagerTests.cpp:101]: Can we also add the same huge-value coverage for `kill-duration` with a `kill-token`? `Retry-After` and `kill-duration` use the same parser, but `kill-duration` goes through `addToken()`, so a direct test would cover that expiry path too.

File: tests/unittests/OfflineStorageTests_SQLite.cpp [hunk 18]
```diff
@@ -183,6 +183,54 @@ TEST_F(OfflineStorageTests_SQLite, DeletedRecordsAreNotReturned)
     EXPECT_THAT(consumer.records[0].id, StrEq("guid2"));
 }
 
+TEST_F(OfflineStorageTests_SQLite, DeleteRecordsRejectsTenantTokenSqlInjection)
```
Comment (lalitb) [inline tests/unittests/OfflineStorageTests_SQLite.cpp:186]: Can we add a couple more fail-closed cases here for the new `DeleteRecords` behavior, like an empty filter and an invalid numeric filter such as `retry_count = "0 OR 1=1"`? Those are now explicitly handled by the code but not covered by this regression test.

File: lib/offline/KillSwitchManager.hpp [hunk 19]
```diff
@@ -156,6 +167,101 @@ namespace MAT_NS_BEGIN {
...
+                if (value[i] < '0' || value[i] > '9')
+                {
+                    return false;
+                }
+            }
+            try
+            {
+                // Only std::out_of_range can fire now (the substring is all digits);
+                // catching it ignores an over-large value rather than crashing.
+                const long long parsed = std::stoll(value.substr(begin, end - begin));
```
Comment (lalitb) [inline lib/offline/KillSwitchManager.hpp:227]: This could be optimized with `std::from_chars` in C++17 or later, but since this project is C++11 and this only runs while parsing response headers, the current `std::stoll` approach looks reasonable to me.

**Review by lalitb — APPROVED** [review]
LGTM. Nice work. I left a few non-blocking comments, but I think it would be good to confirm the tenant token length limit before merging.

File: lib/offline/KillSwitchManager.hpp (hunk 15 above)
Comment (bmehta001) [reply lib/offline/KillSwitchManager.hpp]: I have messaged the Collector team to ask - I was just keeping a large upper bound bc my token is 74 bytes, but I don't actually know if very long tokens could be possible.

File: tests/unittests/CMakeLists.txt (hunk 16 above)
Comment (bmehta001) [reply tests/unittests/CMakeLists.txt:30]: Added KillSwitchManagerTests.cpp to the Android native test list and the iOS unittests xcodeproj

File: tests/unittests/KillSwitchManagerTests.cpp (hunk 17 above)
Comment (bmehta001) [reply tests/unittests/KillSwitchManagerTests.cpp:101]: Added handleResponse_HugeKillDuration_IsClampedNotWrapped (clamped via the shared parser, applied per-token through addToken, token stays blocked) and handleResponse_OutOfRangeKillDuration_DoesNotThrowAndDoesNotBlock to cover the kill-duration/addToken expiry path.

File: tests/unittests/OfflineStorageTests_SQLite.cpp (hunk 18 above)
Comment (bmehta001) [reply tests/unittests/OfflineStorageTests_SQLite.cpp:186]: Added DeleteRecordsEmptyFilterDeletesNothing (empty filter -> no predicate -> nothing deleted) and DeleteRecordsInvalidNumericFilterDeletesNothing (retry_count = '0 OR 1=1' -> rejected as an invalid numeric value, not coerced to 0)

File: lib/offline/KillSwitchManager.hpp (hunk 19 above)
Comment (bmehta001) [reply lib/offline/KillSwitchManager.hpp:227]: Thanks, will keep that in mind if we want to update to C++ 17 in the future.

File: lib/offline/KillSwitchManager.hpp (hunk 15 above)
Comment (bmehta001) [reply lib/offline/KillSwitchManager.hpp]: I haven't heard back from the Aria team, so given the urgency of this PR, I can just drop the size filter.

File: lib/offline/KillSwitchManager.hpp [hunk 20]
```diff
@@ -64,14 +67,22 @@ namespace MAT_NS_BEGIN {
                         // Strip suffix and assume ':all' events of that tenant are killed
                         token.erase(pos, token.length() - pos);
                     }
```
[the remark this answers is not in the document]
Comment (bmehta001) [reply lib/offline/KillSwitchManager.hpp:71]: Fixed in 2b35a2bf — simplified to `token.erase(pos)` (the 1-arg overload erases from `pos` to the end, equivalent to `token.erase(pos, token.length() - pos)`).

File: lib/offline/KillSwitchManager.hpp [hunk 21]
```diff
@@ -62,7 +65,15 @@ namespace MAT_NS_BEGIN {
                     if (pos != std::string::npos)
                     {
                         // Strip suffix and assume ':all' events of that tenant are killed
-                        token.erase(pos, token.length() - pos);
+                        token.erase(pos);
+                    }
```
[the remark this answers is not in the document]
Comment (bmehta001) [reply lib/offline/KillSwitchManager.hpp:71]: Fixed (commit 0b502151, now on the branch via the main-merge `f986d21d`). Dropped the `std::move`: since `it` is a `const_iterator`, `it->second` is a `const std::string&` and `std::move` of a const lvalue binds to the copy constructor — so it copied while implying a move. It's now a plain `std::string token = it->second;` with a comment noting the copy.

File: lib/offline/KillSwitchManager.hpp [hunk 22]
```diff
@@ -156,6 +167,103 @@ namespace MAT_NS_BEGIN {
...
+        // token), and it keeps the value safe if it ever reaches a log/display sink.
+        // No maximum length is enforced: there is no authoritative max token size, and
+        // capping it could let an over-long-but-legitimate tenant escape the kill.
+        static bool isValidTenantToken(const std::string& token)
+        {
+            if (token.empty())
+            {
+                return false;
+            }
+            for (unsigned char c : token)
```
[the remark this answers is not in the document]
Comment (bmehta001) [reply lib/offline/KillSwitchManager.hpp]: Good catch — that was stale. After removing the unconfirmed 256-byte cap, `isValidTenantToken()` no longer enforces any maximum length, so I've updated the PR description: it now says kill-tokens are rejected when empty or containing control characters (no maximum length is enforced, as there's no authoritative max token size). The code and description are now consistent.

File: lib/offline/KillSwitchManager.hpp [hunk 23]
```diff
@@ -156,6 +169,103 @@ namespace MAT_NS_BEGIN {
...
+            {
+                if (value[i] < '0' || value[i] > '9')
+                {
+                    return false;
+                }
+            }
+            try
+            {
+                // Only std::out_of_range can fire now (the substring is all digits);
+                // catching it ignores an over-large value rather than crashing.
```
[the remark this answers is not in the document]
Comment (bmehta001) [reply lib/offline/KillSwitchManager.hpp]: Relaxed the comment in f498d7d3. You're right — `substr()` can also throw (e.g. `std::bad_alloc`); the comment now notes that `std::stoll` itself can only throw `std::out_of_range` here (all-digit substring) but `substr()` may throw too, and either way the `std::exception` catch below ignores the value rather than crashing.

File: tests/unittests/KillSwitchManagerTests.cpp [hunk 24]
```diff
@@ -0,0 +1,197 @@
...
+{
+    // A token containing control characters (e.g. CR/LF) is unsafe -- it could
+    // enable log injection -- and must be ignored.
+    KillSwitchManager manager;
+    HttpHeaders headers;
+    headers.add("kill-tokens", std::string("tenant\r\ninjected"));
+    headers.add("kill-duration", "300");
+    ASSERT_NO_THROW(manager.handleResponse(headers));
+    ASSERT_FALSE(manager.isActive());
+}
```
[the remark this answers is not in the document]
Comment (bmehta001) [reply tests/unittests/KillSwitchManagerTests.cpp:143]: Good suggestion — added `handleResponse_EmptyKillToken_IsRejected` in f498d7d3. It covers both a directly-empty kill-token and a `":all"`-only token (which strips to empty after suffix removal), asserting the manager stays inactive and doesn't block the empty string. Updated the PR description's test count accordingly (18).

File: lib/offline/KillSwitchManager.hpp (hunk 15 above)
Comment (bmehta001) [reply lib/offline/KillSwitchManager.hpp]: Heard back from the Aria team that all tenant tokens are the same size (~74 chars; see `HttpRequestEncoder.cpp:144` "Tenants tokens are usually 74 chars long"), so using 256-byte as a safe ceiling.

File: lib/offline/KillSwitchManager.hpp [hunk 25]
```diff
@@ -156,6 +169,109 @@ namespace MAT_NS_BEGIN {
...
+        // We also bound the length: real tenant tokens are a fixed, small size
+        // (~74 chars; see HttpRequestEncoder.cpp), so kMaxTenantTokenBytes (256) is a
+        // generous ceiling -- it caps memory growth from a malicious/oversized
+        // kill-token in an untrusted HTTP response, while staying far above any real
+        // token so a legitimate tenant can never be made unkillable by this check.
+        static constexpr size_t kMaxTenantTokenBytes = 256;
+        static bool isValidTenantToken(const std::string& token)
+        {
+            if (token.empty() || token.size() > kMaxTenantTokenBytes)
+            {
```
[the remark this answers is not in the document]
Comment (bmehta001) [reply lib/offline/KillSwitchManager.hpp:260]: The description is aligned in the latest commits: it no longer says "no maximum length" -- it now states kill-tokens are rejected when longer than 256 bytes (`kMaxTenantTokenBytes`, a generous ceiling well above the ~74-char real token size), and the `KillSwitchManagerTests` count/coverage were updated for the two boundary tests (a 256-byte token stays killable, a 257-byte token is ignored). Code and description are consistent.

File: tests/unittests/OfflineStorageTests_SQLite.cpp [hunk 26]
```diff
@@ -183,6 +183,120 @@ TEST_F(OfflineStorageTests_SQLite, DeletedRecordsAreNotReturned)
...
+{
+    initializeStorage();
+    ASSERT_THAT(offlineStorage->StoreRecord({"guid1", "tokenA", EventLatency_Normal, EventPersistence_Normal, 1, {}}), true);
+    ASSERT_THAT(offlineStorage->StoreRecord({"guid2", "tokenB", EventLatency_Normal, EventPersistence_Normal, 1, {}}), true);
+
+    // A malicious kill-token value attempting stacked SQL (OR 1=1 / DROP TABLE).
+    // It must be treated as a literal tenant_token value that matches nothing.
+    const std::string malicious = "x\" OR 1=1; DROP TABLE events; --";
```
[the remark this answers is not in the document]
Comment (bmehta001) [reply tests/unittests/OfflineStorageTests_SQLite.cpp]: Fixed in 3c664fb7 -- the test comment now calls the injected value a `tenant_token` (the offline-storage filter field), not a kill-token.

**Review by baijumeswani — APPROVED** [review]
